### PR TITLE
update the functions of math_kernel

### PR DESCRIPTION
**Pull request by haozhihan**

### CODE REVIEW

File: source/module_hsolver/include/math_kernel.h [hunk 1]
```diff
@@ -7,29 +7,249 @@
...
-      const std::complex<FPTYPE>* psi_L, 
-      const std::complex<FPTYPE>* psi_R, 
-      const bool reduce = true);
+struct zdot_real_op
+{
+    FPTYPE operator()(const Device* d,
+                      const int& dim,
+                      const std::complex<FPTYPE>* psi_L,
+                      const std::complex<FPTYPE>* psi_R,
+                      const bool reduce = true);
```
Comment (denghuilu) [inline source/module_hsolver/include/math_kernel.h]:
```suggestion
struct zdot_real_op {
  FPTYPE operator() (
      const Device* d,
      const int& dim,
      const std::complex<FPTYPE>* psi_L,
      const std::complex<FPTYPE>* psi_R,
      const bool reduce = true);
```

Comment (denghuilu) [reply source/module_hsolver/include/math_kernel.h]: Do not change the original format!

File: source/module_hsolver/src/math_kernel.cpp [hunk 2]
```diff
@@ -3,32 +3,186 @@
...
 // Explicitly instantiate functors for the types of functor registered.
 template struct zdot_real_op<double, psi::DEVICE_CPU>;
-}
\ No newline at end of file
+template struct vector_div_constant_op<double, psi::DEVICE_CPU>;
+template struct vector_mul_vector_op<double, psi::DEVICE_CPU>;
+template struct vector_div_vector_op<double, psi::DEVICE_CPU>;
+template struct constantvector_addORsub_constantVector_op<double, psi::DEVICE_CPU>;
+
+// 由于blas将两种精度的接口做了区分，所以我们也得区分。
```
Comment (denghuilu) [inline source/module_hsolver/src/math_kernel.cpp:181]: Use English comments instead

File: source/module_hsolver/include/math_kernel.h (hunk 1 above)
Comment (haozhihan) [reply source/module_hsolver/include/math_kernel.h]: fine!

File: source/module_hsolver/src/math_kernel.cpp (hunk 2 above)
Comment (haozhihan) [reply source/module_hsolver/src/math_kernel.cpp:181]: ok

File: source/module_hsolver/src/math_kernel.cpp [hunk 3]
```diff
@@ -3,32 +3,186 @@
...
+        // qianrui modify 2021-3-14
+        // Note that  ddot_(2*dim,a,1,b,1) = REAL( zdotc_(dim,a,1,b,1) )
+        const FPTYPE* pL = reinterpret_cast<const FPTYPE*>(psi_L);
+        const FPTYPE* pR = reinterpret_cast<const FPTYPE*>(psi_R);
+        FPTYPE result = BlasConnector::dot(2 * dim, pL, 1, pR, 1);
+        if (reduce)
+        {
+            Parallel_Reduce::reduce_double_pool(result);
+        }
+        return result;
```
Comment (denghuilu) [inline source/module_hsolver/src/math_kernel.cpp]:
```suggestion
struct hsolver::zdot_real_op<FPTYPE, psi::DEVICE_CPU> {
  FPTYPE operator() (
      const psi::DEVICE_CPU* d,
      const int& dim,
      const std::complex<FPTYPE>* psi_L,
      const std::complex<FPTYPE>* psi_R,
      const bool reduce) 
  {
    //<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
    // qianrui modify 2021-3-14
    // Note that  ddot_(2*dim,a,1,b,1) = REAL( zdotc_(dim,a,1,b,1) )
    const FPTYPE* pL = reinterpret_cast<const FPTYPE*>(psi_L);
    const FPTYPE* pR = reinterpret_cast<const FPTYPE*>(psi_R);
    FPTYPE result = BlasConnector::dot(2 * dim, pL, 1, pR, 1);
    if (reduce) {
      Parallel_Reduce::reduce_double_pool(result);
    }
    return result;
  }
```
Do not change the original format!

File: source/module_hsolver/src/cuda/math_kernel.cu [hunk 4]
```diff
@@ -30,7 +107,172 @@ FPTYPE zdot_real_op<FPTYPE, psi::DEVICE_GPU>::operator()(
...
+
+
 // Explicitly instantiate functors for the types of functor registered.
 template struct zdot_real_op<double, psi::DEVICE_GPU>;
+template struct vector_div_constant_op<double, psi::DEVICE_GPU>;
+template struct vector_mul_vector_op<double, psi::DEVICE_GPU>;
+template struct vector_div_vector_op<double, psi::DEVICE_GPU>;
+template struct constantvector_addORsub_constantVector_op<double, psi::DEVICE_GPU>;
+
+// 由于blas将两种精度的接口做了区分，所以我们也得区分。
```
Comment (denghuilu) [inline source/module_hsolver/src/cuda/math_kernel.cu:271]: Use English instead.

Comment (haozhihan) [reply source/module_hsolver/src/cuda/math_kernel.cu:271]: [OK]

**Pull request merged**